src/AppInstallerRepositoryCore/RepositorySource.cpp

```diff
@@ -34,52 +34,127 @@ namespace AppInstaller::Repository
             return ISourceFactory::GetForType(details.Type)->Create(details);
         }
 
+        std::chrono::milliseconds GetMillisecondsToWait(std::chrono::seconds retryAfter, size_t randomMultiplier = 1)
+        {
+            if (retryAfter != 0s)
+            {
+                return std::chrono::duration_cast<std::chrono::milliseconds>(retryAfter);
+            }
+            else
+            {
+                // Add a bit of randomness to the retry wait time
+                std::default_random_engine randomEngine(std::random_device{}());
+                std::uniform_int_distribution<long long> distribution(2000, 10000);
+
+                return std::chrono::milliseconds(distribution(randomEngine) * randomMultiplier);
+            }
+        }
+
+        bool IsUpdateSuppressed(const SourceDetails& details)
+        {
+            return std::chrono::system_clock::now() < details.DoNotUpdateBefore;
+        }
+
+        struct AddOrUpdateResult
+        {
+            bool UpdateChecked = false;
+            bool MetadataWritten = false;
+        };
+
         template <typename MemberFunc>
-        bool AddOrUpdateFromDetails(SourceDetails& details, MemberFunc member, IProgressCallback& progress)
+        AddOrUpdateResult AddOrUpdateFromDetails(SourceDetails& details, MemberFunc member, IProgressCallback& progress)
         {
-            bool result = false;
+            AddOrUpdateResult result;
+
+            // Do not update if we are still before the update block time.
+            if (IsUpdateSuppressed(details))
+            {
+                AICLI_LOG(Repo, Info, << "Update is suppressed until: " << details.DoNotUpdateBefore);
+                return {};
+            }
+
             auto factory = ISourceFactory::GetForType(details.Type);
 
+            // If we are instructed to wait longer than this, just fail rather than retrying.
+            constexpr std::chrono::seconds maximumWaitTimeAllowed = 60s;
+            std::chrono::seconds waitSecondsForRetry = 0s;
+
             // Attempt; if it fails, wait a short time and retry.
             try
             {
-                result = (factory.get()->*member)(details, progress);
-                if (result)
+                result.UpdateChecked = (factory.get()->*member)(details, progress);
+                if (result.UpdateChecked)
                 {
                     details.LastUpdateTime = std::chrono::system_clock::now();
+                    result.MetadataWritten = true;

```

Comment:
On success, we should probably zero out `DoNotUpdateBefore` to prevent odd scenarios (unless we are trying to support them).  For instance, if the 429/503 response said "wait 2 days", then a manual update attempt succeeded an hour later, would we still want to continue waiting 2 days for background updates?